### DISCUSSION TITLE
Variable target numbers

### DISCUSSION THREAD

**Discussion by jllancaster** [HighDiceRoller/icepool#181] [Q&A] [answered]
Hi there, wondering if I might be able to get some help here. I am trying to figure out how to code up a mechanic with variable target numbers for the dice pool, but my feeble brain could use some tips on getting started.

1. Roll a pool of 2d6, and a pool of 6d6.
2. Drop lowest or highest from 2d6 pool.
3. Every result in the 6d6 pool that is greater than or equal to the remaining die in the 2d6 pool is considered a success; however, 1s always count as zero successes, and 6s always count as 2 successes.

Thank you very much!

I've made it this far:
```
from icepool import d6

# drop highest (advantage)
target = d6.pool(2)[-1:]

# Remember, since we are dealing with multisets here, the `>=` operator means `issuperset`.
@multiset_function
def var_target(a):
    return a >= target

print(var_target(d6.pool(6)))
```

**Comment by HighDiceRoller** [comment, marked as answer]
Happy to get your message! In this case it's easiest to use the `@map_function` decorator, which runs a function over all outcomes of dice and all tuples-of-outcomes of pools.

```py
from icepool import d6, map_function

@map_function
def var_target(pool, target):
    target = max(target, 2)
    return sum(x >= target for x in pool) + pool.count(6)

output(var_target(d6.pool(6), d6.highest(2, 1)))
```
[You can try this in your browser here.](https://highdiceroller.github.io/icepool/apps/icecup.html?p=eq&c=GYJw9gtgBAlgxgUwA5jAG1hFIAuUAmAbADRQQCGSA+sAK4B2cOMY9AUGwAIXV2POs2+BMCgA3ciCo5JAcwQ4AFCnSkZIeTgCUALjZQDUdZqgBeMuQAei4wtIAmLfsMgFtEPSgBnWhEWWoAD5zWzxgMBAoAJhPFTQtKABqKDiAOjgwBiVCJzZMnCRaJQkpUMUiVLjFHNIKgAsYWTqELyV7UgBGLW62IA)

**Reply by jllancaster** [reply]
Thank you so much, this is awesome and super elegant! I think from here I can incorporate two successes on a roll of 6, and dropping the 1s from the pool. Ultimately, the results from this should be evaluated against an opposed pool of nd6, perhaps canceling successes made in the first roll (haven't figured that out yet). Nice work on Icepool BTW!

Oh actually, perhaps you cancel the 1s here?
`target = max(target, 2)`

And it's so damned elegant I didn't catch you have the doubled 6s in there as well...so nevermind on that!

**Reply by jllancaster** [reply]
Hi there, I seem to have incorporated an opposed pool of d6 here, with each success cancelling a success from the pool. Is there perhaps a better way to go about this?

```
# method to incorporate opposed roll (in-progress)

from icepool import d6, map_function

@map_function
def var_target(pool, target, opposed):
    target = max(target, 2)
    opp_tn = 3
    return (sum(x >= target for x in pool) + pool.count(6)) - sum(x >= opp_tn for x in opposed)

# disadvantage
output(var_target(d6.pool(6), d6.highest(2, 1), d6.pool(6)))
# normal
output(var_target(d6.pool(6), d6, d6.pool(6)))
# advantage
output(var_target(d6.pool(6), d6.lowest(2, 1), d6.pool(6)))
```

**Reply by HighDiceRoller** [reply]
Your code is at least workable, and if you wanted to use the same randomized TN for both pools it would be my preferred approach.

If the two pools have independent TNs, then you alternatively could just use the original function and subtract outside the function call:

```py
from icepool import d6, map_function

@map_function
def var_target(pool, target):
    target = max(target, 2)
    return sum(x >= target for x in pool) + pool.count(6)

# If opposition only gets one success on a 6.
output(var_target(d6.pool(6), d6) - 6 @ (d6 >= 3))
# If opposition gets double successes on 6s like the primary roller does.
output(var_target(d6.pool(6), d6) - var_target(d6.pool(6), 3))
```

**Reply by jllancaster** [reply]
Thank you so much! That is a much-much faster solution than what I posted. Really appreciate it.

**Reply by HighDiceRoller** [reply]
Ah right, the efficiency aspect. The `@map_function` interface is relatively easy to use, but since it expands pools into all possible sorted sequences, it is often not the most efficient approach. An even faster but slightly more complex alternative would be to only map one variable at a time, and furthermore compute the result of a single pool die and use the `@` operator rather than doing a full pool expansion:

```py
from icepool import d6

def var_target(pool_size, target):
    def inner(target):
        pool_die = d6.map(lambda x: (x == 6) + (x >= max(2, target)))
        return pool_size @ pool_die
    return target.map(inner)

output(var_target(6, d6.highest(2, 1)))
```

**Reply by jllancaster** [reply]
Hi there, I am wondering how I might capture the number of successes from var_target(), and use that for the number of dice rolled for the opposed roll?

```
from icepool import d6, map_function

@map_function
def var_target(pool, target):
    target = max(target, 2)
    return sum(x >= target for x in pool) + pool.count(6)

# If opposition only gets one success on a 6.
output(var_target(d6.pool(6), d6) - 6 @ (d6 >= 3))

```
Thanks!

**Reply by HighDiceRoller** [reply]
`var_target(d6.pool(6), d6)` is the number of initial hits; in this case, the opponent has a chance to block each such hit with a 3+ (or whatever), similar to armour saves in Warhammer. We can use the `@` operator again, which means "roll the left side, then roll that many of the right side and sum". For the right side we can use chance of *not* saving each hit:

```py
from icepool import d6, map_function

@map_function
def var_target(pool, target):
    target = max(target, 2)
    return sum(x >= target for x in pool) + pool.count(6)

output(var_target(d6.pool(6), d6) @ (d6 < 3))
```

**Reply by jllancaster** [reply]
That is pretty slick, thank you so much!
